Support programming of Arduino devices in serial passthrough mode

SteveCEvans opened the pull request:
See https://github.com/betaflight/betaflight/issues/5128

SteveCEvans commented:
See https://github.com/betaflight/betaflight/pull/5129

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -936,12 +953,23 @@ static void cliSerialPassthrough(char *cmdline)
+        	// When programming Arduino based devices such as MinimOSD, the DTR line is used to control
+        	// the reset. This parameter is the pin number. For example, 56 (0x38) is port C8. This allows
+        	// any GPIO to be used for driving DTR.
+            dtrPin = atoi(tok);
```
Pin name parsing (`strToPin`) should be used (supplying `C8` as parameter)

SteveCEvans replied:
Excellent! Thanks for that.

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -970,13 +998,21 @@ static void cliSerialPassthrough(char *cmdline)
+
+        // Set the baud rate
+        // Needed as arduino bootloaders may run at 57600 baud
+        serialSetBaudRate(passThroughPort, baud);
```
This should be called only if user supplied new baudrate

SteveCEvans replied:
Noted. Fixed.

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -970,13 +998,21 @@ static void cliSerialPassthrough(char *cmdline)
     }
 
+    // Register control line state callback
+    serialSetCtrlLineStateCb(cliPort, cbCtrlLine);
```
Register only if requested by used

SteveCEvans commented on src/main/interface/cli.c:
```diff
@@ -908,6 +908,23 @@ static void cliSerial(char *cmdline)
 }
 
 #ifndef SKIP_SERIAL_PASSTHROUGH
+static uint32_t dtrPin = 0;
```
Noted and fixed.

SteveCEvans commented on src/main/interface/cli.c:
```diff
@@ -936,12 +953,23 @@ static void cliSerialPassthrough(char *cmdline)
             if (strstr(tok, "tx") || strstr(tok, "TX"))
                 mode |= MODE_TX;
             break;
+#ifdef IOGetByTag
```
Removed

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -1041,7 +1041,9 @@ static void cliSerialPassthrough(char *cmdline)
 
     // Register control line state callback
-    serialSetCtrlLineStateCb(cliPort, cbCtrlLine);
+    if (serialPassthroughDtrPin != IO_NONE) {
```
(personally, I prefer `if (serialPassthroughDtrPin)` - `IO_t` is guarantied to work in bool context)

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -936,12 +977,28 @@ static void cliSerialPassthrough(char *cmdline)
+                	cliPrintLine("Invalid DTR pin");
+                	return;
+                } else {
+                	serialPassthroughDtrPin = IOGetByTag(serialPassthroughDtrTag);
```
`IOGetByTag` does check if pin is available, but will correctly accept `IO_TAG_NONE` - it will validate `strToPin`

```
if (!strToPin(tok, &serialPassthroughDtrTag) || !IOGetByTag(serialPassthroughDtrTag)) {

jflyper commented:
I would just assign one "DTR pin" and use it whenever defined, provided no other serial pass through apps doesn't fiddle with DTR.

PINIO can also be used (is DTR positive or negative?)

SteveCEvans commented:
@jflyper Perhaps a variable could be defined per uart to define the port/pin to be used for DTR on each? The 'serialpassthrough' command would then revert to it's previous definition. Also, each defined DTR line could be pulsed on startup to reset any attached device, and then driven to a negated state. If you like that idea, let me know and I'll implement it. First I need to figure out why the targets-group-rest TARGET fails to build on SITL.

mikeller commented on src/main/interface/cli.c:
```diff
@@ -908,6 +908,46 @@ static void cliSerial(char *cmdline)
+// Use 'serialpassthrough 5 57600 rxtx c8' and then use Ardino to program MinimOSD
+// Use 'serialpassthrough 5 115200' and then use MWOSD configurator to setup MinimOSD
+//
+static void cbCtrlLine(uint16_t data)
```
`cli.c` isn't really the place for this. How about `drivers/serial.c`?

ledvinap replied:
The pin is only used in cli, it's not serial related ...

mikeller replied:
To my understanding it is used for serial passthrough, which is initiated in CLI, but I would not consider it to be part of CLI. Maybe it needs a new file, like the ESC passthrough...

ledvinap replied:
`serialPassthrough()` is in io/serial.c ..
moving cbCtrlLine + serialPassthroughDtrPin will either pollute global namespace or will need interface functions ...

mikeller replied:
Could / should `serialPassthrough()` take care of handling the DTR pin, and should the config be added to this call?

ledvinap replied:
Not sure ... the code is simpler with CLI handling it ...

mikeller replied:
Potentially simpler, but not nicely structured. In terms of maintenance, having all of the serial passthrough code in one place will be easier.

SteveCEvans replied:
I've taken the time to look at the serial driver structure and it doesn't really lend itself to the addition of DTR handling for a couple reasons. Because the openSerialPort() call takes parameters taken from the serialPortConfig_t structure rather than simply a pointer to the structure, expansion of it's parameters to include a DTR pin requires modification of every call to openSerialPort() even if most drivers won't use DTR. I think it simpler to consider the DTR pin usage to be a sideband signal, not really part of the serial port, but rather a reset line controlled by USB.

mikeller replied:
It should not go into the `driver` layer, but into `io`, where `serialPassthrough()` already is.

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -936,6 +977,21 @@ static void cliSerialPassthrough(char *cmdline)
+            // When programming Arduino based devices such as MinimOSD, the DTR line is used to control
+            // the reset. This parameter is the pin name, for example C8. This allows
+            // any GPIO to be used for driving DTR.
+            if (strToPin(tok, &serialPassthroughDtrTag)) {
```
Invalid pin format will be silently ignored

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -936,6 +977,21 @@ static void cliSerialPassthrough(char *cmdline)
+                    cliPrintLine("Invalid DTR pin");
+                    return;
+                } else {
+                    serialPassthroughDtrPin = IOGetByTag(serialPassthroughDtrTag);
```
`IOGetByTag` may still return NULL (pin not supported on target)

ledvinap commented:
@jflyper : DTR is usable only in serialpassthrough, but there may be more arduino-bootloader devices (openLog comes to mind), so single pin may not be enough. Using as implemented in this PR is IMO reasonable compromise.

PINIO is alternative, preferably by passing PINIO index as argument to serialpassthrough. serialpassthrough can simply steal PINIO (no scheduler running). This will also force pin as output outside serialpassthrough, helping with noise-caused reset.

SteveCEvans commented:
Having read the above I think the best thing is to acknowledge that DTR is part of the serial port. I'll define an array for DTR pin mappings for each UART. Then I'll pulse the DTR on each on startup to neatly reset any attached Arduino device and leave it driven high. Finally I'll have the CLI code register a handler for DTR state changes and call a handler in the serial driver to set the DTR state.

mikeller commented:
This looks good now, thanks @SteveCEvans. Don't worry too much about the RAM overflows for now - we'll look into what we can do to make it fit after 3.3 has been released.

SteveCEvans commented:
@mikeller Thanks. Do you have somewhere in the Wiki I could document using this feature outside of the current Wiki, so that can be included at the same time as the feature?

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -970,6 +1003,16 @@ static void cliSerialPassthrough(char *cmdline)
+        } else {
+			// Register the right side baud rate setting routine with the left side which allows setting of the UART
+			// baud rate over USB without setting it using the serialpassthrough command
+			serialSetBaudRateCb(cliPort, serialSetBaudRate, passThroughPort);
```
This may alter baudrate asynchronously to serial transmission, possibly corrupting data
(Actually not a big problem in practice)

SteveCEvans replied:
Indeed, this is a possibility, but this is typically done when a port is opened, before data transmission, not during, and whilst the serial data could be queued in a buffer, the baud rate change isn't buffered. So, trying to change the baud rate after sending some serial data would be dangerous, but doing so beforehand there won't be a race condition.

ledvinap commented on src/main/vcp/usb_prop.h:
```diff
@@ -31,7 +31,7 @@
 /* Includes ------------------------------------------------------------------*/
 /* Exported types ------------------------------------------------------------*/
-typedef struct {
+typedef struct __attribute__ ((packed)) {
```
Is is really necessary? There may be cleaner way to get correct `sizeof`

SteveCEvans replied:
Yes, this is the correct thing to do if matching a data structure to data on the wire. The alternative is to define a set of constant offsets into a block of data which isn't so friendly.

ledvinap commented on src/main/vcp_hal/usbd_cdc_interface.c:
```diff
@@ -187,6 +195,12 @@ static int8_t CDC_Itf_Control (uint8_t cmd, uint8_t* pbuf, uint16_t length)
     LineCoding.format     = pbuf[4];
     LineCoding.paritytype = pbuf[5];
     LineCoding.datatype   = pbuf[6];
+    // If a callback is provided, tell the upper driver of changes in baud rate
```
The structure is just build here - just test that `length == 7` (pbuf[6] is last byte)

SteveCEvans replied:
Yuk! No, it should be the sizeof the structure and the use of the packed attribute facilitates this.

ledvinap replied:
`USBD_CDC_LineCodingTypeDef LineCoding` is used here, that structure is actually not packed ... 
And the structure is rebuilt  byte-by-byte, it's size and structure is defined by unpack code above, not `struct` declaration ...
When casting received packet to the structure it will make sense, but alignment and endianness is problem then

Also note that `length` is actually 8 here(if the code is working, IMO USB is using 16bit units a lot ...

ledvinap replied:
Actually it seems that this code path is simply untested ...

ledvinap commented on src/main/vcp_hal/usbd_cdc_interface.c:
```diff
@@ -187,6 +195,12 @@ static int8_t CDC_Itf_Control (uint8_t cmd, uint8_t* pbuf, uint16_t length)
     LineCoding.paritytype = pbuf[5];
     LineCoding.datatype   = pbuf[6];
+    // If a callback is provided, tell the upper driver of changes in baud rate
+	 if (pbuf && (length == sizeof (LineCoding))) {
```
`pbuf` was already dereferenced, too late to check it ...

SteveCEvans replied:
I'll move the check to cover the read of pbuf.

ledvinap commented on src/main/vcp_hal/usbd_cdc_interface.c:
```diff
@@ -201,7 +215,12 @@ static int8_t CDC_Itf_Control (uint8_t cmd, uint8_t* pbuf, uint16_t length)
   case CDC_SET_CONTROL_LINE_STATE:
-    /* Add your code here */
+    // If a callback is provided, tell the upper driver of changes in DTR/RTS state
+    if (pbuf && (length == sizeof (uint16_t))) {
```
It would be cleaner to build aligned linestate here. Also this code ignores USB endianness

SteveCEvans replied:
You mean to interpret DTR/RTS encoding here? Endianness isn't about to change. We could add a load of endian aware macros around all code processing data to/from the wire, but we're not about to change endianness, so is this really worth the effort?

ledvinap replied:
No, just implement is the same as `LineCoding`:
```
LineState = (uint16_t)(pbuf[0] | (pbuf[1] << 8);
```

ledvinap replied:
A lot of IO code in BF is endianness aware ... In this case it's mainly to match `LineCoding`

ledvinap commented on src/main/vcpf4/usbd_cdc_vcp.c:
```diff
@@ -124,6 +131,12 @@ static uint16_t VCP_Ctrl(uint32_t Cmd, uint8_t* Buf, uint32_t Len)
       //Note - hw flow control on UART 1-3 and 6 only
       case SET_LINE_CODING:
+          // If a callback is provided, tell the upper driver of changes in baud rate
+     	 if (Buf && (Len == sizeof (*plc))) {
```
(Ok, ST isn't that pedantic about alignment and endianness)
`Len >= sizeof (*plc)` may be safer test (from assert above)...

SteveCEvans replied:
If Len is greater than sizeof (*plc) then we're probably not looking at a LINE_CODING structure!

ledvinap replied:
I don't know (or want to study) USB standard in this regard. Someone written assert above that way, checking that enough data are available. Host USB implementation that sends too big wLength will cause pretty nasty bug ...

ledvinap commented on src/main/drivers/serial.c:
```diff
@@ -86,6 +86,51 @@ void serialSetMode(serialPort_t *instance, portMode_e mode)
     instance->vTable->setMode(instance, mode);
 }
 
+void serialSetCtrlLineStateCb(serialPort_t *serialPort, void (*cb)(serialPort_t *context, uint16_t ctrlLineState), serialPort_t *context)
```
This is getting a bit out of hand. Using something like posix `ioctl` (with vararg) (http://man7.org/linux/man-pages/man2/ioctl.2.html) or `setsockopt` (pasing cmd, ptr, size) vill reduce number of virtual functions.

Driver can switch on CMD, it will be compact and easier to read ...

SteveCEvans replied:
Agreed that we could go down the ioctl route, however that's not suitable for registering callbacks. This isn't the normal kernel/user split. These callbacks are being called in USB ISR context.

ledvinap replied:
Yes, but exporting so many virtual functions feels bad. Only registration should be done through `setsockopt`-like call, driver will invoke callback directly ...

SteveCEvans commented on src/main/interface/cli.c:
```diff
@@ -970,6 +1003,16 @@ static void cliSerialPassthrough(char *cmdline)
             serialSetMode(passThroughPort, mode);
         }
+
+        // Set the baud rate if specified otherwise allow to be set via USB
```
@ledvinap The baud rate isn't passed from cliSerialPassthrough() into serialPassthrough(). I could add that parameter and then move the call to either serialSetBaudRate() or serialSetBaudRateCb() if that would be better.

Note sure I understand the above comment. What do you mean by the "baud rate used by device"? Do you mean as set in the Ports tab of the GUI? Unfortunately the syntax of the serialpassthrough command doesn't currently lend itself to specifying a port number for DTR and specifying or not, the baud rate. We can easily code any behaviour we want here, but we may need to change the serialpassthrough command semantics to do so.

ledvinap replied:
The idea was something like:
- baud < 1 - (optional) register USB baud hook, start with `abs(baud)`
- baud < 0  - register USB baud hook, start with inherited value
- baud == 0 - use inherited value, no change over USB
- baud > 0 - use baudrate,  no change over USB

ledvinap replied:
Old behavior was to keep baudrate if port is open, request it otherwise

SteveCEvans replied:
@ledvinap Now that the SERIAL_DTR resource is being used to define the pin, the serialpassthrough command has reverted to it's original param list. If the baud rate is specified that will be used as a fixed value, but if it is omitted then the baud rate can be set over USB.

ledvinap replied:
Well, someone will eventually find combination where setting speed from USB is undesirable (SBUS testing on linux comes to mind), but we can handle it later ...

ledvinap replied:
BTW: It would be nice to allow baudrate forwarding even when opening fresh port - for example by specifying initial baudrate as negative.

SteveCEvans replied:
Alternatively the behaviour of displaying "closed, specify baud" could be changed to set a default baud rate (say 57600) and then allow the rate to be changed over USB. Much easier for the end user than documenting setting baud to -1 etc.

ledvinap replied:
Yes, that may do, but message about default + USB change shall be printed ..

SteveCEvans commented on src/main/drivers/serial.c:
```diff
@@ -86,6 +86,51 @@ void serialSetMode(serialPort_t *instance, portMode_e mode)
+	UNUSED(serialPort);
+	UNUSED(ioTagDtr);
+#else /* UNIT_TEST */
+	serialPinConfigMutable()->ioTagDtr[serialPort->identifier] = ioTagDtr;
```
It is? I'd not realised that. If so, that's excellent news as it provides a mechanism to save the DTR port assignment for a UART once and subsequently not have to specify it. I'll look into this some more. :)

SteveCEvans replied:
@ledvinap Are you sure? pgResetFn_serialPinConfig is called at startup and initialises the pin config from the serialDefaultPin array. Looking at this has made me realise that I should really be employing the resources infrastructure here to define the DTR pins. If I define a new SERIAL_DTR function, then if I can assign instance 1 to use port B10, for example, then I won't need to specify the DTR port with the serialpassthrough command and can also ensure that all SERIAL_DTR resources are correctly initialised, so no floating outputs!

SteveCEvans replied:
Unfortunately it looks like the resource command won't let me add a function to the resource table, just edit the port assignments of those which are already present.

ledvinap replied:
That's only reset function, it's called explicitly by user ... 

Other pins are (IIRC) configured using `cli set`, at least for softserial, you can mimic the code ... 

I'm still not that sure about adding generic DTR support just to allow arduino programming ...

SteveCEvans commented on src/main/drivers/serial.c:
```diff
@@ -86,6 +86,51 @@ void serialSetMode(serialPort_t *instance, portMode_e mode)
+	UNUSED(serialPort);
+	UNUSED(ctrlLineState);
+#else /* UNIT_TEST */
+	ioTag_t serialPassthroughDtrPin = serialPinConfig()->ioTagDtr[serialPort->identifier];
```
Yes, serialSetCtrlLineState() is only called after serialSetCtrlLineStateDtrPin() which initialises the pin. This order is guaranteed by:

```serialSetCtrlLineStateDtrPin(right, serialPassthroughDtrTag);
serialSetCtrlLineStateCb(left, cbCtrlLine, (void *)right);```

ledvinap replied:
It is a bit more complicated when you take persistent storage into account. In current PR serialSetCtrlLineStateDtrPin() is used before registering callback, but it's far from obvious from `drivers/serial.h` interface

mikeller commented:
@SteveCEvans: Documentation should be added to the relevant paragraph in `docs/Serial.md`. We can then link this document from the wiki.

jflyper commented:
I think these would be enough for limited purpose of triggering reset during serial pass through.
1. Expand `serialPassthrough` to receive `pinio` resource number.
2. The `serialPassthrough` registers DTR callback that triggers `pinio`.

ledvinap commented:
@jflyper : I like this minimalist implementation too

- Another `serialPassthrough` related function may register it DTR forwarding and baudrate manipulation, avoiding another argument and keeping serialPassthrough clean/generic ... 

- PINIO-BOX may be used. Adding BOXTRUE will handle it (and support some other nice features in future)

- When used is programming OSD over FC, we can expect a bit of competence ...

SteveCEvans commented:
This is a all a compromise between usability, architectural elegance and me coming up to speed with your architecture. I can see the merit in removing the setCtrlLineStateDtrPin() routine if we're to use a generic "PINIO" resource, however, we have the opportunity here to define a SERIAL_DTR resource for not much more effort so that a UART's IO will be defined by its SERIAL_TX, SERIAL_RX and SERIAL_DTR resources. This would enable the use of any IO pin as DTR for any UART. Flexible and easily understood. Usability is *really* important with this kind of thing.

If we can simply document the "resource" CLI incantation to associate a GPIO pin as the DTR line of a given UART in a non-volatile manner, we can then ensure it's always driven correctly on reset, and we can then ignore the DTR completely on the "serialpassthrough" command as DTR would then be configured automatically.

To extend things further I note that the MWOSD configurator has a passthrough checkbox which then lets you select a UART. Not sure how this selection gets passed over USB (yet), but this might enable us to use the "serialpassthrough" command with no parameters and then select UART, and baud rate from the configurator.

SteveCEvans commented:
DTR is now defined as a resource as per the example below from a REVO.

```# resource
resource MOTOR 1 B00
resource MOTOR 2 B01
resource MOTOR 3 A03
resource MOTOR 4 A02
resource MOTOR 5 A01
resource MOTOR 6 A00
resource PPM 1 B14
resource PWM 1 B14
resource PWM 2 B15
resource PWM 3 C06
resource PWM 4 C07
resource PWM 5 C08
resource PWM 6 C09
resource LED_STRIP 1 A01
resource SERIAL_TX 1 A09
resource SERIAL_TX 3 B10
resource SERIAL_TX 4 A00
resource SERIAL_TX 6 C06
resource SERIAL_RX 1 A10
resource SERIAL_RX 3 B11
resource SERIAL_RX 4 A01
resource SERIAL_RX 6 C07
resource SERIAL_DTR 6 C08
resource INVERTER 1 C00
resource I2C_SCL 1 B08
resource I2C_SDA 1 B09
resource LED 1 B05
resource LED 2 B04
resource SPI_SCK 1 A05
resource SPI_SCK 3 C10
resource SPI_MISO 1 A06
resource SPI_MISO 3 C11
resource SPI_MOSI 1 A07
resource SPI_MOSI 3 C12
resource ESCSERIAL 1 B14
resource ADC_BATT 1 C02
resource ADC_CURR 1 C01```

The driving of DTR is then completely automatic. As baud rate can now also be controlled by the MW_OSD configurator, the serialpassthrough command just requires a single argument specifying the serial port to use. Note that MW_OSD must be built with MAX_SOFTRESET defined or it falls over when DTR is asserted.

ledvinap commented on src/main/io/serial.c:
```diff
@@ -412,6 +412,18 @@ serialPort_t *openSerialPort(
+    ioTag_t ioDtrTag = serialPinConfig()->ioTagDtr[identifier];
+
+    // Initialise DTR pin if defined
+    if (ioDtrTag != IO_TAG_NONE) {
```
IO was designed to support  `if (ioDtrTag) {`

SteveCEvans commented:
@mikeller I've incorporated the majority of the review feedback. I've stuck with the serialSetCtrlLineStateCb()/serialSetCtrlLineState()/serialSetBaudRateCb() functions rather than add some kind of IOCTL support. I've also added a SERIAL_DTR resource rather than use PINIO as I think it's much neater and more intuitive. Finally I've added documentation to serial.md. Ready to merge I think bar the memory space issues on some targets that you've already noted.

mikeller commented:
@SteveCEvans: Thanks for this. I am still humming and harring about having this converted to pinio. It seems to be a match, and using a generic GPIO abstraction for this will improve the maintainability with respect to hardware abstraction and adding support for new MCU families. @jflyper: What are your thoughts?

Also, please note that this will only be merged once 3.3 has been released, so there is still time to work on it if needed.

ledvinap commented on src/main/io/serial.c:
```diff
@@ -430,6 +446,18 @@ void closeSerialPort(serialPort_t *serialPort)
+    // Negate DTR pin if defined
+    if (ioDtrTag) {
+        IO_t ioDtr = IOGetByTag(ioDtrTag);
+        IOWrite(ioDtr, CTRL_LINE_STATE_DTR);
```
1/true shall be passed instaed of `CTRL_LINE_STATE_DTR`

ledvinap commented on src/main/rx/nrf24_kn.c:
```diff
@@ -86,27 +86,27 @@ static void switch_channel(void)
 
 static void decode_bind_packet(uint8_t *packet)
 {
```
Please don't include (unrelated) whitespace changes, it makes history and review complicated. Separate PR is necessary

ledvinap commented on src/main/target/REVO/target.h:
```diff
@@ -208,6 +208,7 @@
 #define USE_UART6
 #define UART6_RX_PIN            PC7
 #define UART6_TX_PIN            PC6
+#define UART6_DTR_PIN           PC8
```
Is this safe default for all users?

ledvinap commented on src/main/vcp_hal/usbd_cdc_interface.c:
```diff
@@ -182,26 +190,40 @@ static int8_t CDC_Itf_Control (uint8_t cmd, uint8_t* pbuf, uint16_t length)
-    LineCoding.format     = pbuf[4];
-    LineCoding.paritytype = pbuf[5];
-    LineCoding.datatype   = pbuf[6];
+    if (pbuf && (length == sizeof (LineCoding))) {
```
I still think that `LineCoding` as local information about line coding and `CDC_SET_LINE_CODING` format are mixed. Checking `sizeof(LineCoding)` make sense if received buffer is cast to `USBD_CDC_LineCodingTypeDef` to parse it. But coding is done byte-by-byte, so both can be totally unrelated ...

ledvinap replied:
Also `USBD_CDC_LineCodingTypeDef` is unpacked, so this test will fail ...

ledvinap replied:
unpacked `USBD_CDC_LineCodingTypeDef` is still used for length check ...

ledvinap commented on src/main/vcpf4/usbd_cdc_vcp.c:
```diff
@@ -102,9 +109,8 @@ void ust_cpy(LINE_CODING* plc2, const LINE_CODING* plc1)
  * @param  Len: Number of data to be sent (in bytes)
  * @retval Result of the opeartion (USBD_OK in all cases)
  */
-static uint16_t VCP_Ctrl(uint32_t Cmd, uint8_t* Buf, uint32_t Len)
```
(is it better to keep ST style here)

SteveCEvans commented:
@ledvinap @mikeller I'm not clear on what is outstanding on this PR.

mikeller commented:
@SteveCEvans: As I stated before, this adds a lot of change, and will require considerable effert to be maintained in the future, for a relatively small target audience. I think it will be better to do this with pinio, which will result in considerably less change, at the cost of slightly more complex setup.

SteveCEvans commented:
@mikeller Given that the USB code would still be required, and we'd still need to hook this into driving an IO line, via PINIO vs a SERIAL_DTR specified resource, I don't see a big difference in complexity. Can you please explain how PIOIO should be used in such a case. Thanks.

mikeller commented:
@SteveCEvans: See @jflyper's comment https://github.com/betaflight/betaflight/pull/5129#issuecomment-366511347.

ledvinap commented on src/main/io/serial.c:
```diff
@@ -571,13 +571,17 @@ void serialPassthrough(serialPort_t *left, serialPort_t *right, serialConsumer *
             LED0_ON;
             uint8_t c = serialRead(left);
+            // Make sure there is space in the tx buffer
+            while (!serialTxBytesFree(right));
```
(This will only postpone overflow until both serial buffers are full)

SteveCEvans replied:
@ledvinap This prevents more data being written to the tx buffer until there is space. The buffer will be getting drained by the UART driver so this loop will only stall execution for one byte transmit time (dependant on baud rate). This change is needed because the size of the UART buffers has been reduced which was breaking avrdude.

I'm just completing a switch from specifying SERIAL_DTR resources to using PINIO resources as requested in your review. I'll be honest; I don't like using PINIO as it's less intuitive (and ease of use is really important), but if that's what you guys want I'll do that.

ledvinap replied:
Yes, it will effectively double buffer size - both receive and transmit buffers are used.

ledvinap commented on docs/Serial.md:
````diff
@@ -97,6 +97,77 @@ For example. If you have your MWOSD connected to UART 2, you could enable commun
+
+```
+# resource
+resource MOTOR 1 B00
````
Quite long ... keeping only important lines (+ context) is better

SteveCEvans replied:
OK, reduced to just serial resources.

ledvinap commented on src/main/drivers/serial.h:
```diff
@@ -48,6 +48,10 @@ typedef enum {
 } portOptions_e;
 
+// Define known line control states which may be passed up by underlying serial driver callback
+#define CTRL_LINE_STATE_DTR 1
```
```
#define CTRL_LINE_STATE_DTR (1<<0)
#define CTRL_LINE_STATE_RTS (1<<1)

SteveCEvans replied:
OK

ledvinap replied:
Seems that the change got lost ...

ledvinap commented on src/main/rx/rx.c:
```diff
@@ -421,6 +421,8 @@ bool rxUpdateCheck(timeUs_t currentTimeUs, timeDelta_t currentDeltaTime)
             rxIsInFailsafeMode = (frameStatus & RX_FRAME_FAILSAFE) != 0;
             rxSignalReceived = !rxIsInFailsafeMode;
             needRxSignalBefore = currentTimeUs + needRxSignalMaxDelayUs;
+    } else if (cmpTimeUs(currentTimeUs, rxNextUpdateAtUs) > 0) {
```
Unrelated to this PR

SteveCEvans replied:
Fixed

ledvinap commented on src/main/vcp_hal/usbd_cdc_interface.h:
```diff
@@ -65,6 +65,18 @@
+/* The following structures groups all needed parameters to be configured for the
+   ComPort. These parameters can modified on the fly by the host through CDC class
+   command class requests. */
+typedef struct __attribute__ ((packed))
```
This struct is available in HAL, with different name. But it's not a big problem ...

SteveCEvans replied:
USBD_CDC_LineCodingTypeDef isn't packed which is the point here.

ledvinap commented:
> I'm just completing a switch from specifying SERIAL_DTR resources to using PINIO resources as requested in your review. I'll be honest; I don't like using PINIO as it's less intuitive (and ease of use is really important), but if that's what you guys want I'll do that.

If DTR was useful in more places, it would be ok. But only arduino programming is using it, but code impact is quite noticeable.
PINIO will allow similar functionality and can be localized only into seriallpassthrough. It will also provide features like inversion for free ...

ledvinap commented on docs/Serial.md:
```diff
@@ -89,14 +89,64 @@ Cleanflight can enter a special passthrough mode whereby it passes serial data t
 
 To initiate passthrough mode, use the CLI command `serialpassthrough` This command takes three arguments.
```
up to four arguments

ledvinap commented on docs/Serial.md:
````diff
@@ -89,14 +89,64 @@ Cleanflight can enter a special passthrough mode whereby it passes serial data t
+To assign the DTR line to another pin use the following command.
+
+```
+# resource PINIO 1 c05
````
`#` is comment in cli, it may confuse users

SteveCEvans replied:
OK. Fixed

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -955,6 +956,9 @@ static void cliSerialPassthrough(char *cmdline)
                 mode |= MODE_TX;
             break;
+        case 3:
+            pinioDtr = atoi(tok);
```
Maybe check that pinio is valid.

SteveCEvans replied:
OK

ledvinap commented on src/main/target/SITL/target.c:
```diff
@@ -537,3 +537,9 @@ void spektrumBind(rxConfig_t *rxConfig)
     UNUSED(rxConfig);
     printf("spektrumBind");
 }
+
```
Not needed

SteveCEvans replied:
OK

ledvinap commented on src/main/io/serial.c:
```diff
@@ -511,12 +512,19 @@ static void nopConsumer(uint8_t data)
     UNUSED(data);
 }
 
+static void cbCtrlLine(void *context, uint16_t ctrl)
```
This can be moved back to cli ... this is not strictly passthrough related (and baudrate callback is set there too)

SteveCEvans replied:
OK. Thanks for the feedback.

jflyper commented:
I'm very sorry to say this at this late in the development stage, but I think this entire approach is way overkill; it has 350 lines of change spanning 20 files just to trigger a pin up and down in response to DTR changes during serial passthrough, which is an edge case.

Why not completely separate DTR transition detection + pin up/down manipulation from serial code?

This of course imposes certain procedure to be taken when flashing Arduinos:

- Enable DTR to pinio: "set pinio_dtr = 1". This will activate DTR callback. Reboot.
- Do serial passthrough operation.
- (Optionally) Disable DTR to pinio: "set pinio_dtr = -1".

SteveCEvans commented:
OK, I'll strip this back to basics then, as per 649cd04, but using PINIO. Complexity started creeping in at 4b06f8c as a result of review.

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -927,6 +927,15 @@ static void cliSerial(char *cmdline)
+{
+    int pinioDtr = (int)(long)context;
+
+    pinioSet(pinioDtr, ~ctrl & CTRL_LINE_STATE_DTR);
```
`pinioSet(pinioDtr, !(ctrl & CTRL_LINE_STATE_DTR));` is easier to understand ...

SteveCEvans replied:
OK

ledvinap commented on src/main/interface/cli.c:
```diff
@@ -998,6 +1025,13 @@ static void cliSerialPassthrough(char *cmdline)
+#ifdef USE_PINIO
+    // Register control line state callback
+    if (pinioDtr) {
+        serialSetCtrlLineStateCb(cliPort, cbCtrlLine, (void *)(long)(pinioDtr - 1));
```
`(intptr_t)` is correct type here

SteveCEvans replied:
OK

SteveCEvans replied:
OK

ledvinap commented:
@jflyper : Most of the changes are documentation and USB driver augmentation.
Actual impact to the code is quite small.
With ioctl-like mechanism for serial ports (it is needed anyway, but in separate PR), it will have negligible impact. And it provides facilities for future use.
For example GPS passthrough now supports baudrate over USB, making some operations much easier.

DTR (ans USB signaling) may be used for other uses too ...

ledvinap commented:
@SteveCEvans : Rebase/quash would be necessary before merging, but wait for @jflyper ...

SteveCEvans commented:
@ledvinap @jflyper @mikeller Unless I've lost the plot completely I now believe all your inputs to have been addressed.

richard-scott commented:
I see that this has been merged into 3.4.x on build [685](https://ci.betaflight.tech/job/Betaflight/685/changes).

This has stopped my AnyFC F7 from booting. I get the serial port show up in Windows, but I can't connect. Build 684 is ok, but anything 685 or above doesn't allow me to connect to it via USB.

I'm not yet sure if the FC is functional as I don't have a battery with me right now.

richard-scott commented:
Logging from Arming on the bench looks odd.
[LOG00001.zip](https://github.com/betaflight/betaflight/files/1835072/LOG00001.zip)

mikeller commented:
@richard-scott: Can you open an issue for this problem please?

richard-scott commented:
@mikeller looks like I've been beaten to it: https://github.com/betaflight/betaflight/issues/5496. I've added comments to that issue, but happy to open a dedicated one if you feel it's required?

richard-scott commented:
I've opened a new [issue ](https://github.com/betaflight/betaflight/issues/5502)as #5496 isn't the root of my issue.

conkerkh commented on lib/main/STM32_USB_Device_Library/Class/cdc/src/usbd_cdc_core.c:
```diff
@@ -547,7 +547,7 @@ uint8_t  usbd_cdc_Setup (void  *pdev,
       {
         /* Transfer the command to the interface layer */
-        APP_FOPS.pIf_Ctrl(req->bRequest, NULL, 0);
+        APP_FOPS.pIf_Ctrl(req->bRequest, (uint8_t *)&req->wValue, sizeof(req->wValue));
```
I suppose this uses Control EP of CDC Interface?

ledvinap replied:
Seems that CDC control is transferred over EP0 ...
